pkg/common/cluster_actions.go

```diff
@@ -557,6 +566,14 @@ type RemoveClientRoleAction struct {
 	Msg      string
 }
 
+type SetServiceAccountRolesAction struct {
+	ClientName string
+	Ref        *v1alpha1.KeycloakClient
+	Roles      []string
+	Realm      string
+	Msg        string
+}
```

Comment:
Can't we perhaps use [existing actions](https://github.com/keycloak/keycloak-operator/blob/ac6619083c859e5843b7fc4d76e100d14b3d096f/pkg/common/cluster_actions.go#L530-L558)?

Reply:
I thought about that too. However, we need to:
* get the user (by name)
* add set the roles
* update the users

Now, we also need to evaluate the name of the user as you suggested. That is a composite action, and I am not sure the code will be simpler if is will be split into different (existing) actions, passing along state.

So I modeled it into a single action, for the client to implement.